source/extensions/tracers/xray/tracer.cc

```diff
@@ -44,17 +145,59 @@ Tracing::SpanPtr Tracer::startSpan(const std::string& span_name, const std::stri
       span_ptr->setSampled(true);
       break;
     case SamplingDecision::NotSampled:
-      span_ptr->setSampled(false);
+      // should never get here. If the header has Sampled=0 then we never call startSpan().
+      RELEASE_ASSERT(false, "X-Ray startSpan called despite HTTP header having sampled=0");
       break;
     default:
       break;
     }
   } else {
-    span_ptr->setTraceId(Hex::uint64ToHex(ticks));
+    span_ptr->setTraceId(generateTraceId(time_source_.systemTime()));
   }
   return span_ptr;
 }
 
+XRay::SpanPtr Tracer::createNonSampledSpan() const {
+  auto span_ptr = std::make_unique<XRay::Span>(time_source_, *daemon_broker_);
+  const auto ticks = time_source_.monotonicTime().time_since_epoch().count();
+  span_ptr->setId(ticks);
+  span_ptr->setName("NotSampled");
+  span_ptr->setTraceId(generateTraceId(time_source_.systemTime()));
+  span_ptr->setSampled(false);
+  return span_ptr;
+}
+
+void Span::setTag(absl::string_view name, absl::string_view value) {
+  constexpr auto SpanContentLength = "content_length";
+  constexpr auto SpanMethod = "method";
+  constexpr auto SpanStatus = "status";
+  constexpr auto SpanUserAgent = "user_agent";
+  constexpr auto SpanUrl = "url";
+  constexpr auto HttpUrl = "http.url";
+  constexpr auto HttpMethod = "http.method";
+  constexpr auto HttpStatusCode = "http.status_code";
+  constexpr auto HttpUserAgent = "user_agent";
+  constexpr auto HttpResponseSize = "response_size";
```

Comment:
it seems surprising to me that these are local to a particular function. Shouldn't they be declared in a place that the consumers of the tag can reference?

Or, if these are meant for sending to a non-OSS system, maybe a comment pointing to the doc enumerating the spans would be good.

Also I would probably not use `auto` here; I'd probably use:
```
  constexpr char SpanContentLength[] = "content_length";
```
etc

As I'm not even sure what the `auto` will resolve to in this case.

Reply:
I've added a comment pointing to the documentation.

>  `constexpr char SpanContentLength[] = "content_length";`

clang-tidy complains "Avoid C-style arrays" when I use that notation, which I think is absurd in this context. But I also hate the warning showing up all over my files.

> As I'm not even sure what the auto will resolve to in this case.

`constexpr auto SpanContentLength ...;` is equivalent to `constexpr char* SpanContentLength ...` which is equivalent to `const char * const SpanContentLength ...`